### DISCUSSION TITLE
What happened to this emulator? :(

### DISCUSSION THREAD

**Discussion by jrmoore** [visualboyadvance-m/visualboyadvance-m#1322] [General]
I reckon it's been a while since I last used it, it must have been around the time the MFC interface was being deprecated, maybe 10 years or more ago. I came to VBA-M directly since it is what I was using, but things have changed quite a lot. I have tried different builds, both (stable?) releases and nightly ones, but I still haven't found one in which things look and work as expected. I'm using Windows, in case that matters, although I suppose not anymore with it being cross platform.

For example, Alt+F4 doesn't seem to work to close the emulator on the current release when a ROM is being emulated, and not at all in the current nightly. In that nightly the FAudio thing seems ignore the emulation speed being choppy at the same time, but there are alternatives. However, one can't use the number keys to input values, either the ones in the top row or the numpad, it's as if those keystrokes are intercepted somehow, therefore configuration that calls for keyboard input is broken. I even got a crash when switching to search for a 16-bit value instead of an 8-bit one in the "Find cheat" section in one of the recent versions...

The UI is inconsistent regardless of the build: different fonts depending on the control or label, misaligned controls all around, dialogs have handles for resizing but they can't be resized, some are sized to fit all of the controls that any of their tabs have while the "About" one resizes itself, outdated links (support forum leads to GitHub, translations to the main Transifex, not the project), capitalization is all over the place, pixel filters affect the position and size of the OSD display, double clicking an item in the cheat lists to edit it leads to a crash... I don't know, I don't remember it being this way before.

With wxWidgets the thing went from a ~4MB binary to a 10x one and that is less responsive, I don't think many people are using old machines incapable of running the emulator these days, but it saddens me to see things this way, at least it's not XAML levels of slow. It's just I would have thought the new UI (wxWidgets, vs. MFC) to be more polished, flawless even, since its code wouldn't be as old. I don't know if it's the price to pay for having it be the same for several platforms, but I have fond memories of the GTK port as well, I think even the SDL one!

So I thought, okay, perhaps I can track down some of these bugs / annoyances, tackle them and send PRs or something, at least the UI stuff should be some resource descriptions like it was in MFC and not straight up code à la Win32's CreateDialog(), and that led me to clone the repo since I already had Visual Studio installed. But it seems the main development is not being done on Windows, or not with VS, because there are no project files anymore by default, even the build instructions are to go with CMake and Ninja. I know I'm out of touch with the current way of doing things, but damn, things seemed easier back then: recursive clone, open solution, hack in.

Setting up a dev environment for this is already quite involved, going for elevation and running PS scripts for vcpkg unless a user root is used instead, is it really more flexible compared to having dependencies be tagged submodules (for example)? Not that I want to be building wxWidgets on every build... it's just that I see enough black box dependencies already (from the perspective of not knowing anything of the project or its internals), I was wondering if things could be streamlined (it's certainly possible that vcpkg is already simplifying things, and I just need to learn to like it and used to it).

I was able to get projects for VS by not specifying Ninja as the generator in the CMake invocation, to at least have some visibility on it all, to see some structure (there is no proper documentation, right?). I could see no output using Ninja within VS, and with the projects I get to see warnings (and errors) at least, the linker reports there being unreachable code in the end, so there seems to be more things that need looking at underneath the UI.

By the way, are CMake generators equivalent? I mean, are the build / link options the same regardless of the generator used? Because I could see the projects in VS linking release builds to debug OpenAL libraries by default, in case something needed tweaking in that CMakeLists file or something. Not to the same extent, but CMake reminded me of the craziness libtool introduced in projects, certainly better than some pure makefile-based build system's I've seen though.

PS. An MFC build using the current internals isn't possible anymore, is it? I noticed there being an MFC branch still, but it being 1000s of commits behind master I suppose getting it up to speed would be a waste of time, if it were even possible. What was the last MFC build, the 2.0.0 beta2 listed in SourceForge? To think I was initially looking to playing a couple of games and not delving so deep in the whole thing :D

**Comment by Squall-Leonhart** [comment]
cmake-gui will get you a standard vcproj and solution set up, modern visual studio can open cmake and configure itself without having to

F-audio (if and when it eventually works properly) will be a xaudio style interface layered on sdl that doesn't have the framerate hitching caused by sdl2 not having a stable audio mixer.

**Comment by jrmoore** [comment]
Thanks! That GUI showed more CMake variables that influence things, even if not directly related to VBA-M.

I still don't know why, but configuring with CMake gets Git to try to update the vcpkg root as if it were a repo but fails if it's not. I was using the current release of vcpkg instead of cloning that project and no `.git` dir inside lead to it trying to clone the repo there. I wanted to use a fixed version of anything involved instead of rolling releases not to introduce more chaos, at least to get things started...

Another thing I realized is that even if one disables most stuff (FFmpeg, SDL, link, ...) when configuring with CMake, the vcpkg invocation still downloads those dependencies with CMake itself looks for pkgconf. A NuGet binary is also added to the build dir, for... gettext? I understand it isn't used for anything else.

FAudio seems to be the only one in master that interferes with the emulation speed, but since it isn't the default and wasn't shipped in any release (I think), it's okay. There is an occasional chirp with all of them when the mute audio is option is chosen for turbo, the audio accelerates before being muted, only a minor tweak maybe. The turbo % seems to be erratic though, the status bar reports it matching the setting at low % (2-4x), but that setting is already broken at 5x speed (status bar reads 30x).

<hr>

The revision that introduced the problem with keyboard inputs in the emulator GUI was [d543784](https://github.com/visualboyadvance-m/visualboyadvance-m/commit/d543784a3d93bc6b9802f6dd74ff36b5b0b93833) on May 5th, since then the UI controls except the input configuration don't accept any input. There's a note in that commit message about widgets needing to use the `VBAM_USER_INPUT_EVENT` event since then, so perhaps that change was not committed yet. The Alt+F4 breakage was introduced earlier, but I haven't dissected that yet.

<hr>

Answering something I asked earlier, it seems something does need to be changed somewhere in CMake definitions, while Zlib is detected correctly, different libraries depending on whether the build is debug or release based, OpenAL (and something else iirc) links to debug libs even in release builds:

```
-- Found ZLIB: optimized;D:/Code/vcpkg/installed/x64-windows-static/lib/zlib.lib;debug;D:/Code/vcpkg/installed/x64-windows-static/debug/lib/zlibd.lib (found version "1.3.1")
-- Found OpenGL: opengl32
-- Looking for gettext
-- Looking for gettext - found
The source specified has already been added to the list of available package sources. Provide a unique source.
-- Found Gettext: D:/Code/VBA-M-WIP/build/nuget/Gettext.Tools.0.22.5/tools/bin/msgmerge.exe (found version "0.22.5")
-- Found OpenAL: D:/Code/vcpkg/installed/x64-windows-static/debug/lib/OpenAL32.lib
```

Either that, or the CMake invocation I used for that Ninja build was wrong (or something to do with my laptop, ... no clue honestly, all of this is foreign to me and convoluted (for now perhaps)):

```
cmake .. -DCMAKE_TOOLCHAIN_FILE=D:/Code/vcpkg/scripts/buildsystems/vcpkg.cmake -DVCPKG_TARGET_TRIPLET=x64-windows-static \
-DENABLE_DEBUGGER=OFF -DENABLE_LINK=OFF -DENABLE_FFMPEG=OFF -DENABLE_ONLINEUPDATES=OFF -DENABLE_GBA_LOGGING=OFF \
-DENABLE_DIRECT3D=OFF -DENABLE_FAUDIO=OFF -DENABLE_SDL=OFF -DBUILD_TESTING=OFF -DCMAKE_BUILD_TYPE=RelWithDebInfo -G Ninja
```
